Src/FluentAssertions/Equivalency/CyclicReferenceDetector.cs

```diff
@@ -36,7 +36,11 @@ public bool IsCyclicReference(ObjectReference reference, Reason reason = null)
 
             if (reference.IsComplexType)
             {
-                if (orderedReferences.Contains(reference))
+                if (!observedReferencesByReferencedObject.TryGetValue(reference.Object, out var existingReference))
```

Comment:
I actually feel like the way `ObjectReference` implements `Equals` is a bit of a code smell. It's breaking the rules of the `Equals` method. The MSDN documentation explicitly states:

> The following statements must be true for all implementations of the `Equals(Object)` method.
> [..]
> * `x.Equals(y)` returns the same value as `y.Equals(x)`.

Because `ObjectReference` also cares about whether `x` is a parent of `y`, by definition it will never return `true` for both `x.Equals(y)` and `y.Equals(x)`.

I'm not convinced it's actually safe to assume that a `HashSet<T>` would work correctly, because it's not actually defined whether the `HashSet<T>` would be using `dataInHashSet.Equals(callerProvidedValue)` or `callerProvidedValue.Equals(dataInHashSet)` -- and in fact it _could_ do both in different places, because of the documented invariant that they are equivalent.

That's why I renamed the method in the proposed changes.

Reply:
I have added a custom `IEqualityComparer` that ensures that the dictionary uses reference comparisons.

Reply:
> Because `ObjectReference` also cares about whether `x` is a parent of `y`, by definition it will never return `true` for both `x.Equals(y)` and `y.Equals(x)`.

This is by design. The `ObjectReference` exists to detect that two references point to the same object within the same path.

Reply:
I understand that, but it violates a fundamental rule of the .NET Framework. Surely it doesn't decrease confidence in the code to have this method under a different name??

Reply:
Renaming something is fine, but changing the semantics is not.

Reply:
Oh, but how does this change the semantics? I mean, it did before the custom `IEqualityComparer` was put in, but with that in place, doesn't this have exactly the same behaviour as before??

Reply:
> I'm not convinced it's actually safe to assume that a `HashSet<T>` would work correctly, because it's not actually defined whether the `HashSet<T>` would be using `dataInHashSet.Equals(callerProvidedValue)` or `callerProvidedValue.Equals(dataInHashSet)` -- and in fact it _could_ do both in different places, because of the documented invariant that they are equivalent.

Yeah, I agree.

Reply:
Okay, I believe this should be resolved to everyone's satisfaction now :-)

* ObjectReference.cs now uses `Equals` again, but this is okay because `Equals` is now symmetric. This is unit tested.
* CyclicReferenceDetector.cs can then confidently use `HashSet<ObjectReference>`. All existing tests pass.

I had thus far been amending my commit to incorporate PR review, having observed this elsewhere in the repository, but in this case the changes required for this are in a separate commit on the branch so they can be easily inspected independently.